src/goto-programs/remove_calls_nobody.cpp

```diff
@@ -0,0 +1,116 @@
+/*******************************************************************\
+
+Module: Remove calls to functions without a body
+
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Remove calls to functions without a body
+
+#include "remove_calls_nobody.h"
+
+void remove_calls_nobodyt::remove_call_nobody(
+  goto_programt &dest,
+  goto_programt::targett target,
+  const exprt &lhs,
+  const exprt::operandst &arguments)
+{
+  assert(target->is_function_call());
+  assert(!dest.empty());
+
+  goto_programt tmp;
+
+  // evaluate function arguments -- they might have
+  // pointer dereferencing or the like
+  forall_expr(it, arguments)
+  {
+    goto_programt::targett t=tmp.add_instruction();
+    t->make_other(code_expressiont(*it));
+    t->source_location=target->source_location;
+    t->function=target->function;
+  }
+
+  // return value
+  if(lhs.is_not_nil())
+  {
+    side_effect_expr_nondett rhs(lhs.type());
+    rhs.add_source_location()=target->source_location;
+
+    code_assignt code(lhs, rhs);
+    code.add_source_location()=target->source_location;
+
+    goto_programt::targett t=tmp.add_instruction(ASSIGN);
+    t->source_location=target->source_location;
+    t->function=target->function;
+    t->code.swap(code);
+  }
+
+  // kill call
+  target->type=LOCATION;
+  target->code.clear();
+  target++;
+
+  dest.destructive_insert(target, tmp);
+}
+
+void remove_calls_nobodyt::remove_calls_nobody(
+  goto_programt &goto_program,
+  const goto_functionst &goto_functions)
+{
+  for(goto_programt::targett it=goto_program.instructions.begin();
+      it!=goto_program.instructions.end();) // no it++
+  {
+    if(!it->is_function_call())
```

Comment:
I'd be tempted to break the logic of this loop up into two parts:

```C++
bool is_opaque_function_call(const goto_programt::target instruction, const goto_functions &goto_functions)
{
  if(!instruction->is_function_call()) 
    return false;
...
if(!goto_function.body_available())
  return true;
}
```

And then this function becomes:

```C++
for(goto_programt::targett it=goto_program.instructions.begin();
    it!=goto_program.instructions.end();) // no it++
{
  if(is_opaque_function_call(it, goto_function)
  {
    remove_call_nobody(goto_program, it, cfc.lhs(), cfc.arguments());
  }
  else
  {
    it++;
  }
}
```

This way it is easy to see what the loop is doing and when the iterator gets advanced